compiler/src/iree/compiler/Codegen/SPIRV/SPIRVBreakDownLargeVector.cpp

```diff
@@ -99,11 +103,12 @@ struct BreakDownCastExtractExtend final : OpRewritePattern<arith::ExtUIOp> {
         extractOp.getLoc(), extOp.getType(),
         rewriter.getZeroAttr(extOp.getType()));
 
+    // Extract each elements assuming little-endian style encoding.
```

Comment:
Can't comment on an unchanged line:

```
    if (extractSrcType.getNumElements() <= 4)
      return failure();
```

I'm not convinced that this bit extraction is correct for all bit widths and packings, but I think it will work for POT widths. Wdyt about tightening up this condition so that it only allows the cases tested, and leave a TODO/revisit for other bit widths? Maybe add this to our list of debt to evaluate.

Reply:
This is accounted for in the upstream type emulation pass. So we probably want to just use that as a way to do all emulation.

Reply:
> I think it will work for POT widths.

What does "POT" mean here?

> Wdyt about tightening up this condition so that it only allows the cases tested, and leave a TODO/revisit for other bit widths? Maybe add this to our list of debt to evaluate.

> This is accounted for in the upstream type emulation pass. So we probably want to just use that as a way to do all emulation.

Yes we can tighten this pattern for now and take another look later. I added this pattern earlier because the whole picture is rather complicated. To do it in the most proper way, we need to consider a few things:
* How to emulate i4 types itself. In SPIR-V different bitwidth types require different capabilities. i8/i16 is so. i32 is always there; but it would mean that we see i32 -> 8xi4. Then that comes to the issue of:
* How to handle large vectors with more than 4 elements. Need an lowering pattern for vector.bitcast for such cases, which is missing right now. Even after adding it, all following vector ops involving vector<8xi4> needs to be lowered properly. Which comes to--
* How to sequence various vector patterns. There are lots and it's a long standing issue of difficulties to properly organize them together.
* We can try to avoid generating large vectors if possible. But it needs more thinking on how to have various patterns to work together..
* etc.

Reply:
Sorry. Power of two